### PR TITLE
preStart events support

### LINKED ISSUE #4901: preStart lifecyle event support
/kind feature

## Which functionality do you think we should add?

It should be possible to add a preStart event to a component, for odo to define an init container in the deployment to run the specified command.

## Why is this needed?

Running s2i components needs to run a specific init container to copy files to the persistent volume.

### PR DESCRIPTION
**Pull request by feloy**
**What type of PR is this?**

/kind feature

**What does this PR do / why we need it**:

- [x] support of preStart events
- [x] Includes initContainers in component deployment 
- [x] use of specific commit of devfile library

**Which issue(s) this PR fixes**:

Fixes #4901 

**PR acceptance criteria**:

- [x] Unit test 

- [x] Integration test 

- [x] Documentation 

- [x] Update changelog

- [x] I have read the [test guidelines](https://github.com/openshift/odo/blob/master/docs/dev/test-architecture.adoc)

**How to test changes / Special notes to the reviewer**:

- Run `odo create --s2i wildfly wildfly`
- Update the created devfile with the following content
- Run `odo push`
- The component should start correctly

```
commands:
- exec:
    commandLine: /opt/odo/bin/s2i-setup && /opt/odo/bin/assemble-and-restart
    component: s2i-builder
    group:
      isDefault: true
      kind: build
  id: s2i-assemble
- exec:
    commandLine: /opt/odo/bin/run
    component: s2i-builder
    group:
      isDefault: true
      kind: run
  id: s2i-run
- exec:
    commandLine: /opt/odo/bin/run
    component: s2i-builder
    group:
      isDefault: true
      kind: debug
  id: s2i-debug
- apply:
    component: copy-app-root-container
  id: copy-app-root
events:
  preStart:
  - copy-app-root
components:
- container:
    command:
    - /bin/sh
    - -c
    - "cp -R /opt/app-root /mnt/"
    image: docker.io/openshift/wildfly-100-centos7:latest
    volumeMounts:
    - name: app-root
      path: /mnt/app-root
  name: copy-app-root-container
- container:
    endpoints:
    - name: http-8778
      targetPort: 8778
    - name: http-8080
      targetPort: 8080
    env:
    - name: DEBUG_PORT
      value: "5858"
    - name: ODO_S2I_SCRIPTS_URL
      value: /usr/libexec/s2i
    - name: ODO_S2I_SCRIPTS_PROTOCOL
      value: image://
    - name: ODO_S2I_SRC_BIN_PATH
      value: /tmp
    - name: ODO_S2I_DEPLOYMENT_DIR
      value: /deployments
    - name: ODO_S2I_WORKING_DIR
      value: /home/jboss
    - name: ODO_S2I_BUILDER_IMG
      value: centos7/s2i-base-centos7
    - name: ODO_SRC_BACKUP_DIR
      value: /opt/app-root/src-backup
    - name: ODO_S2I_CONVERTED_DEVFILE
      value: "true"
    image: docker.io/openshift/wildfly-100-centos7:latest
    mountSources: true
    sourceMapping: /tmp/projects
    volumeMounts:
    - name: app-root
      path: /opt/app-root
  name: s2i-builder
- name: app-root
  volume:
    size: 1Gi
metadata:
  name: wildfly
  version: 1.0.0
schemaVersion: 2.0.0
```

### CODE REVIEW

**Comment by mohammedzee1000** [comment]
Verified as working
```bash
> mkdir sample
cp -avrf ../odo/tests/examples/source/wildfly/* ./sample
> cd sample
> odo create --s2i wildfly wildfly
Validation
Warning: wildfly is not fully supported by odo, and it is not guaranteed to work
 ✓  Validating component [756ms]
Conversion
 ✓  Successfully generated devfile.yaml and env.yaml for provided S2I component

Please use `odo push` command to create the component with source deployed
> cat devfile.yaml
commands:
- exec:
    commandLine: /opt/odo/bin/s2i-setup && /opt/odo/bin/assemble-and-restart
    component: s2i-builder
    group:
      isDefault: true
      kind: build
  id: s2i-assemble
- exec:
    commandLine: /opt/odo/bin/run
    component: s2i-builder
    group:
      isDefault: true
      kind: run
  id: s2i-run
- exec:
    commandLine: /opt/odo/bin/run
    component: s2i-builder
    group:
      isDefault: true
      kind: debug
  id: s2i-debug
components:
- container:
    endpoints:
    - name: http-8080
      targetPort: 8080
    - name: http-8778
      targetPort: 8778
    env:
    - name: DEBUG_PORT
      value: "5858"
    - name: ODO_S2I_SCRIPTS_URL
      value: /usr/local/s2i
    - name: ODO_S2I_SCRIPTS_PROTOCOL
      value: image://
    - name: ODO_S2I_SRC_BIN_PATH
      value: /tmp
    - name: ODO_S2I_DEPLOYMENT_DIR
      value: /deployments
    - name: ODO_S2I_WORKING_DIR
      value: /home/jboss
    - name: ODO_S2I_BUILDER_IMG
      value: centos7/s2i-base-centos7
    - name: ODO_SRC_BACKUP_DIR
      value: /opt/app-root/src-backup
    - name: ODO_S2I_CONVERTED_DEVFILE
      value: "true"
    image: docker.io/openshift/wildfly-100-centos7:latest
    mountSources: true
    sourceMapping: /tmp/projects
  name: s2i-builder
metadata:
  name: wildfly
  version: 1.0.0
schemaVersion: 2.0.0
> ls
devfile.yaml  pom.xml  src
> vi devfile.yaml
> cat devfile.yaml
commands:
- exec:
    commandLine: /opt/odo/bin/s2i-setup && /opt/odo/bin/assemble-and-restart
    component: s2i-builder
    group:
      isDefault: true
      kind: build
  id: s2i-assemble
- exec:
    commandLine: /opt/odo/bin/run
    component: s2i-builder
    group:
      isDefault: true
      kind: run
  id: s2i-run
- exec:
    commandLine: /opt/odo/bin/run
    component: s2i-builder
    group:
      isDefault: true
      kind: debug
  id: s2i-debug
- apply:
    component: copy-app-root-container
  id: copy-app-root
events:
  preStart:
  - copy-app-root
components:
- container:
    command:
    - /bin/sh
    - -c
    - "cp -R /opt/app-root /mnt/"
    image: docker.io/openshift/wildfly-100-centos7:latest
    volumeMounts:
    - name: app-root
      path: /mnt/app-root
  name: copy-app-root-container
- container:
    endpoints:
    - name: http-8778
      targetPort: 8778
    - name: http-8080
      targetPort: 8080
    env:
    - name: DEBUG_PORT
      value: "5858"
    - name: ODO_S2I_SCRIPTS_URL
      value: /usr/libexec/s2i
    - name: ODO_S2I_SCRIPTS_PROTOCOL
      value: image://
    - name: ODO_S2I_SRC_BIN_PATH
      value: /tmp
    - name: ODO_S2I_DEPLOYMENT_DIR
      value: /deployments
    - name: ODO_S2I_WORKING_DIR
      value: /home/jboss
    - name: ODO_S2I_BUILDER_IMG
      value: centos7/s2i-base-centos7
    - name: ODO_SRC_BACKUP_DIR
      value: /opt/app-root/src-backup
    - name: ODO_S2I_CONVERTED_DEVFILE
      value: "true"
    image: docker.io/openshift/wildfly-100-centos7:latest
    mountSources: true
    sourceMapping: /tmp/projects
    volumeMounts:
    - name: app-root
      path: /opt/app-root
  name: s2i-builder
- name: app-root
  volume:
    size: 1Gi
metadata:
  name: wildfly
  version: 1.0.0
schemaVersion: 2.0.0
> oc new-project myproject
Now using project "myproject" on server "https://api.ci-ln-s3lnirb-f76d1.origin-ci-int-gce.dev.openshift.com:6443".

You can add applications to this project with the 'new-app' command. For example, try:

    oc new-app centos/ruby-25-centos7~https://github.com/sclorg/ruby-ex.git

to build a new example application in Ruby.
> odo push

Validation
 ✓  Validating the devfile [241359ns]

Updating services
 ✓  Services and Links are in sync with the cluster, no changes are required

Creating Kubernetes resources for component wildfly
 ✓  Added storage app-root to wildfly
 ✓  Waiting for component to start [54s]
 ✓  Waiting for component to start [291ms]

Applying URL changes
 ✓  URL http-8778: http://http-8778-app-myproject.apps.ci-ln-s3lnirb-f76d1.origin-ci-int-gce.dev.openshift.com/ created
 ✓  URL http-8080: http://http-8080-app-myproject.apps.ci-ln-s3lnirb-f76d1.origin-ci-int-gce.dev.openshift.com/ created

Syncing to component wildfly
 ✓  Checking files for pushing [2ms]
 ✓  Syncing files to the component [6s]

Executing devfile commands for component wildfly
 ✓  Waiting for component to start [325ms]
 ✓  Executing s2i-assemble command "/opt/odo/bin/s2i-setup && /opt/odo/bin/assemble-and-restart" [2s]
 ✓  Executing s2i-run command "/opt/odo/bin/run" [3s]

Pushing devfile component "wildfly"
 ✓  Changes successfully pushed to component
> oc get deployments
NAME          READY     UP-TO-DATE   AVAILABLE   AGE
wildfly-app   1/1       1            1           104s
> oc describe deployment wildfly-app
...
  Init Containers:
   copy-app-root-container-copy-app-root-1:
    Image:      docker.io/openshift/wildfly-100-centos7:latest
    Port:       <none>
    Host Port:  <none>
    Command:
      /bin/sh
      -c
      cp -R /opt/app-root /mnt/
    Environment:
      PROJECTS_ROOT:   /projects
      PROJECT_SOURCE:  /projects
    Mounts:
      /mnt/app-root from app-root-wildfly-app-vol (rw)
   copy-supervisord:
    Image:      registry.access.redhat.com/ocp-tools-4/odo-init-container-rhel8:1.1.10
    Port:       <none>
    Host Port:  <none>
    Command:
      /usr/bin/cp
    Args:
      -r
      /opt/odo-init/.
      /opt/odo/
    Environment:  <none>
    Mounts:
      /opt/odo/ from odo-supervisord-shared-data (rw)
...
```

File: pkg/devfile/adapters/kubernetes/component/adapter.go [hunk 1]
```diff
@@ -411,28 +411,44 @@ func (a *Adapter) createOrUpdateComponent(componentExists bool, ei envinfo.EnvSp
 	labels["component"] = componentName
 	labels[componentlabels.ComponentTypeLabel] = componentType
 
-	containers, err := generator.GetContainers(a.Devfile, parsercommon.DevfileOptions{})
+	allContainers, err := generator.GetContainers(a.Devfile, parsercommon.DevfileOptions{})
 	if err != nil {
 		return err
 	}
 
+	var initContainers []corev1.Container
```
Comment (mohammedzee1000) [inline pkg/devfile/adapters/kubernetes/component/adapter.go:419]: the fact that we have to do this speaks of a need for improvement on the library side. We should be able to fetch init containers and the remaining containers seperately. 
Nice one though

**Review by mohammedzee1000 — APPROVED** [review]
Looks good to me
/lgtm

File: pkg/devfile/adapters/kubernetes/component/adapter.go (hunk 1 above)
Comment (feloy) [reply pkg/devfile/adapters/kubernetes/component/adapter.go:419]: Right, the devfile team has made a change on the library for this recently: https://github.com/devfile/library/pull/108

**Comment by mohammedzee1000** [comment]
Timed out
/retest

File: pkg/devfile/validate/events.go [hunk 2]
```diff
@@ -3,13 +3,6 @@ package validate
...
 func validatePreStart(preStart []string) (err error) {
-	// This is odo specific validation. There is still discussion about how PreStart should be implemented.
-	// https://github.com/devfile/api/issues/204
-	// https://github.com/openshift/odo/issues/4187
-	// This is here to prevent anyone from using PreStart event until we have a proper implementation
-	if len(preStart) != 0 {
-		return &UnsupportedFieldError{fieldName: "preStart"}
-	}
 	return nil
 }
```
Comment (valaparthvi) [inline pkg/devfile/validate/events.go:7]: We don't need this now, do we?

Comment (feloy) [reply pkg/devfile/validate/events.go]: Yes, you are right, I removed the associated code

**Comment by valaparthvi** [comment]
@feloy Do we need integration tests to test preStart event support?

**Comment by feloy** [comment]
> @feloy Do we need integration tests to test preStart event support?

I'm not sure, as it is not a normal use case for users. The support will be tested through the S2I conversion using preStart events with the PR https://github.com/openshift/odo/pull/4918

**Comment by dharmit** [comment]
> @feloy Do we need integration tests to test preStart event support?

It's not user facing at the moment. It's not possible to use any odo commands to set/unset `preStart` events in a devfile. So, IMO, it is OK to not have integration tests here.

**Pull request merged**